### ISSUE TITLE
[Feature]: Make Loader Data Global Data

### ISSUE THREAD

**Issue by Stancobridge** [remix-run/remix#771]
### What is the new or updated feature that you are suggesting?

Though I have not used much of Remix, from the documentation, I didn't see a way to access router loader data globally, I don't mean with the use of useFetcher

In react-query, every data is automatically global which you can access using the queryKey if this concept can be adopted in remix it will also be good, since most developers use react-query it will be a comfortable development context for the developers using react-query.

Considering the case:
In my route I have 3 components that use different data for each, so there is a need to make the request in each of the component not only in the main route, if I update data in One Component I want another component in the same route refetch it's data, use the returned response the update request or refetch its contents based on some conditions, I don't want any other component to refetch it's data.

Features I think that can be implemented to help with global server data are:
1. Each Loader can have an optional unique key(this can be the route since every route is unique and also the api route) that identifies the loader: `export const loader = withKey('route-link', async ():  LoaderFunction => {})`
2. useFetcher can use this key to know which API route to fetch and return cache data if already fetch: `useFetcher('route-link')`
3. Cache Response can be configure when to cache and when on a global level or per route
4. useCacheData can be used to access cached data only: useCacheData('route-link')
5. I don't know if useFetcher can be used programmatically no only through d Form Component that it returns

Main Points:
1. Make Loader Global
2. And Also Maker useFetcher programmable

### Why should this feature be included?

giving the developer ability to control the global data from one base point

**Comment by sergiodxa** [comment]
You can use the useMatches hook to access the data of every route currently active, and a few more things too, like the route handle export.

**Comment by Stancobridge** [comment]
> You can use the useMatches hook to access the data of every route currently active, and a few more things too, like the route handle export.

from https://remix.run/docs/en/v1/api/remix#usematches returns the matched routes,

it has not nothing with fetching data in a component

**Comment by sergiodxa** [comment]
I understood you wanted to be able to get the data from the loaders.

If you want to fetch data from any loader you can use useFetcher, that will give you ` fetcher.load` method you can call passing a URL and call the loader of that route, e.g. `fetcher.loader("/my/api/endpoint")`.

If you want that data fetched there be global you can move it to root and store it inside a context provider. Another option is that you can still use RQ/SWR.

But remember, if you move the data fetching to the client with useFetcher or RQ/SWR you will lose the option to do SSR of that component since you are going to fetch data client-side only. And if you don't fetch data using the routes loader Remix will not be able to revalidate after a form submit (you will have to do it yourself, either manually or by triggering RQ/SWR revalidation), you will lose the performance optimizations Remix can do for you like prefetch on link hover, only fetch the data a child route needs instead of refetching data from parents (you may still get this because of the client-side cache tho), a more things I'm not remembering right now.

**Comment by nickjs** [comment]
@sergiodxa I understood the context of the original issue raised here same as you, and indeed, `useMatches()` works great for accessing data from other nested loaders. However, because of the name (and possibly the implementation itself?), eslint will complain about not being allowed to use a React hook inside a non-component function (i.e. `loader`).

I'd like to propose exporting either an alias or separate function (if the implementation needs to be altered to not rely on any other hooks) called `getMatches()` that can be safely invoked on the server.

If you think that's a worthwhile endeavor I'm happy to open a PR to add that. I can also open a separate issue from this one if that would be helpful.

Thank you!

**Comment by sergiodxa** [comment]
@nickjs the thing with accessing data from a loader inside another loader is that they run at the same time so there's no data to access when the loader run.

For client-side navigation Remix will only call each loader as individual requests. So they will also run in parallel, and depending on your deployment provider they can even run in different hardwares making impossible to share data between them without using an external service like Redis.

Another things is that even if Remix is only calling a single nested loader the only way to have access to the parent loaders data would be if Remix sent the whole data in useMatches to every loader call, with a POST request it may be possible, with a GET request like the ones used for loaders it could be too much data to send in the URL and you can't use the body, and loaders can be changed to use POST because they will not be cacheable anymore.

And yes, you can run useMatches inside a non-hook or non-component function because it's a React hook, it needs access to the internal context Remix has.

---

So, if you need the same data in two loaders, you need to query/fetch it in both loaders, if you only need the data in the React component of two routes which are rendered at the same time (parent and child) then you can return the data only from the parent route loader and use the useMatches to read the data on the child route component.

**Comment by sergiodxa** [comment]
@nickjs don't worry! it's not obvious at the first, I have been using Remix for a year now and learnt all of this things.

**Comment by willin** [comment]
> 

how to use root loader data in other pages' loader function?

**Comment by sergiodxa** [comment]
> how to use root loader data in other pages' loader function?

You can’t because they run at the same time, in parallel, not in a serie, so when other loaders run the root loaders is also running.

**Comment by divmgl** [comment]
I'm glad we still have this issue open. I think I'll return to Remix once it has an elegant way of doing this. Last I spoke with @ryanflorence I proposed allowing loaders to pool promises into an array as a second parameter in the loader function so that nested loaders can `await` on them. That way we're not issuing multiple requests for the same data for a single page.

This isn't as relevant in the client-side because you can access this data with `useLoaderData` and `useMatch`, but if you ever want to redirect or bring back data in nested loaders based on the same data, you'll need to fetch the data over and over again.

**Comment by sergiodxa** [comment]
Passing a list of promises from other loaders wouldn’t be reliable.

The only time all the loaders are guaranteed to run all in the same server is on the initial document request, after that child loaders may be called without the parent loaders, or if they ask run together they could even will have different requests, in some platforms they may even have different physical hardware.

If your code expects other loaders to run to get data from there then in those cases it will break.

What you want is to batch your requests/queries, there’s an example showing how to do it using the dataloader package and the express adapter, this way on the initial document request to only query the same endpoint once and the rest keep working.

**Comment by divmgl** [comment]
I'm going to repeat here what I've said on Discord. I understand how Remix _works_; it doesn't mean that the use cases are invalid.

> and two or more endpoint needed to fetch or query the same data you would have to do it on each endpoint, there works be no way to share the request between endpoints without caching them somewhere.

...unless the endpoints were nested somehow or there was middleware, which is very common in backend systems. Almost every API framework I've ever used, even going back to the ASP.NET MVC days, has the concept of middleware.

Here's one use-case. Say I have this route:

```
/admin/profile
```

`admin` has an `<Outlet />`, so it'll render the `profile` route. In order to avoid regular users reaching the `admin` route at all, the `admin` loader fetches some kind of resource to verify the identity of the user. Let's say it uses the JWT token in the cookie provided in the request, and let's assume we don't store the entire user object in the JWT token. The `admin` loader gets the ID from the token, fetches the `User` resource in a data store, and checks the user's role. If the user is not an admin, the `admin` loader redirects. All's well.

However, the `profile` route now needs this same resource again if the user IS an admin. So we fetch the `User` resource again. Normally this wouldn't be a problem, **but then the profile has yet another subroute**:

```
/admin/profile/security
```

...which AGAIN has to fetch the user. At this point we've now fetched the same `User` resource three times.

I would love to just be able to get the `User` that was grabbed from `admin` the one time. A common pattern in other frameworks is to create middleware that will store this user in the context of the request and child routes can use it. Sure, the loaders in Remix were run in parallel 🙄 yes, I get it, I've been told this a thousand times, and I realize this is antithetical to the way Remix works.

But hear me out: if there was some kind of **global state per request**, this problem would be mitigated. We'd store a promise of some kind there (one of the loaders would certainly beat another loader into placing a promise there) and **all of the loaders would be able to await on that promise**. When that promise is finished, all of the other loaders that `await` on it would continue in parallel.

I think there's an argument to be made here for just having plain ol' middleware, but that's a battle for another time :) There's lots of reasons why I think middleware would be useful, but at least having some kind of global state per request would somewhat mitigate this particular issue.

**Comment by sergiodxa** [comment]
I think your problem is not the duplicated fetching, but the duplication of code.

Because even with middleware the data would have been read one per API endpoint. Following your example, if you have a `/admin`, `/admin/profile` and `/admin/profile/security` endpoints in a normal API built using middleware, the code to get the user would have been extracted away from the controllers of those endpoints, but each endpoint would have still run that code and do the fetch for the User resource.

Middleware would not solver duplicated fetching, only duplicated code. You could in Remix create a function to abstract again all of this code and return the user or a 401 or redirect response if not authenticated. You would of course still need to call it yourself on every loader and action, but this is, right now, the most similar thing you will have to middleware without using Express middleware or another server middleware implementation.

But again, middleware would not solve the fact that the User resource is being fetched three times, because they are three individual request so they can't share data. The only scenario where you could avoid the resource being fetched three times would have been on the initial document request where all loaders runs as part of the same HTTP request.

---

That aside, I think there's an issue or discussion here about adding a pre and post hooks functions server side so you could run code before your loaders/actions, which would enable a global middleware support, and you could check the URL yourself to know if you should run or not your code if you want to limit to to some routes.

I agree that it could be useful for code organization, I think the main issue of middleware would be that typing them is way harder.

**Comment by kiliman** [comment]
@divmgl I used to think the same as you. I even added a patch to Remix to include a `X-Remix-Navigation-Id` header to track the same _logical_ request even if they are _multiple_ parallel requests. But it really only works if you have a single server that is guaranteed to process all the requests. So at that point, it's simpler to use a server-side cache that you manage.

Anyway, the initial request can still benefit from a shared context cache. My example adds a cache object to the `getLoadContext` context. Then in my loaders, I call a helper to get a value from cache (and pass a Promise). The first caller will initialize the promise, and subsequent callers (in the same request) will await on that promise.

```ts
export function getFromCache(
  cache: any,
  key: string,
  getter: () => Promise<any>,
) {
  let promise = cache[key]
  if (!promise) {
    promise = getter()
    cache[key] = promise
  }
  return promise
}

// from your loader
// since cache is the same on document request, only one call to getUser will be
// made. The other loaders will await the same promise
const user = await getFromCache(context.cache, 'user', () => getUser(123))
```

https://github.com/kiliman/remix-playground
